expression/constant_propagation.go

```diff
@@ -0,0 +1,235 @@
+func (s *propagateConstantSolver) validPropagateCond(cond Expression, funNameMap map[string]bool) (*Column, *Constant) {
+	if eq, ok := cond.(*ScalarFunction); ok {
+		if _, ok := funNameMap[eq.FuncName.L]; !ok {
+			return nil, nil
+		}
+		if col, colOk := eq.Args[0].(*Column); colOk {
+			if con, conOk := eq.Args[1].(*Constant); conOk {
+				return col, con
+			}
+		}
+		if col, colOk := eq.Args[1].(*Column); colOk {
+			if con, conOk := eq.Args[0].(*Constant); conOk {
+				return col, con
+			}
+		}
+	}
+	return nil, nil
+}
+
+// pickNewEQConds tries to pick new equal conds and puts them to retMapper.
+func (s *propagateConstantSolver) pickNewEQConds(visited []bool) (retMapper map[int]*Constant) {
+	retMapper = make(map[int]*Constant)
+	for i, cond := range s.conditions {
+		if !visited[i] {
+			col, con := s.validPropagateCond(cond, eqFuncNameMap)
+			if col != nil {
+				visited[i] = true
+				if s.tryToUpdateEQList(col, con) {
+					retMapper[s.getColID(col)] = con
+				} else if s.foreverFalse {
+					return
+				}
+			}
+		}
+	}
+	return
+}
+
+// tryToUpdateEQList tries to update the eqList. When the eqList has store this column with a different constant, like
+// a = 1 and a = 2, we set conditions to false.
+func (s *propagateConstantSolver) tryToUpdateEQList(col *Column, con *Constant) bool {
+	if con.Value.IsNull() {
+		s.foreverFalse = true
+		s.conditions = []Expression{&Constant{
+			Value:   types.NewDatum(false),
+			RetType: types.NewFieldType(mysql.TypeTiny),
+		}}
+		return false
+	}
+	id := s.getColID(col)
+	oldCon := s.eqList[id]
+	if oldCon != nil {
+		log.Warnf("old %s new %s", oldCon, con)
+		if !oldCon.Equal(con) {
+			s.foreverFalse = true
+			s.conditions = []Expression{&Constant{
+				Value:   types.NewDatum(false),
+				RetType: types.NewFieldType(mysql.TypeTiny),
+			}}
+		}
+		return false
+	}
+	s.eqList[id] = con
+	return true
+}
+
+func (s *propagateConstantSolver) solve(conditions []Expression) []Expression {
+	var cols []*Column
+	for _, cond := range conditions {
+		s.conditions = append(s.conditions, SplitCNFItems(cond)...)
+		cols = append(cols, ExtractColumns(cond)...)
+	}
+	for _, col := range cols {
+		s.insertCol(col)
+	}
+	if len(s.columns) > MaxPropagateColsCnt {
+		log.Warnf("[const_propagation]Too many columns in a single CNF: the column count is %d, the max count is %d.", len(s.columns), MaxPropagateColsCnt)
+		return conditions
+	}
+	s.propagateEQ()
+	s.propagateInEQ()
+	for i, cond := range s.conditions {
+		if dnf, ok := cond.(*ScalarFunction); ok && dnf.FuncName.L == ast.OrOr {
+			dnfItems := SplitDNFItems(cond)
+			for j, item := range dnfItems {
+				dnfItems[j] = ComposeCNFCondition(PropagateConstant([]Expression{item}))
+			}
+			s.conditions[i] = ComposeDNFCondition(dnfItems)
+		}
+	}
+	return s.conditions
+}
+
+func (s *propagateConstantSolver) getColID(col *Column) int {
+	code := col.HashCode()
```

Comment:
`HashCode` will be invoked in both `getColID` and `insertCol`

can we define a private attribute `hashCode` to avoid it.
function `HashCode` first check whether hashCode is nil,
if it is nil, then calculate hashcode; else return `hashCode` directly.